{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "082ca2da-28b7-42fa-8c70-92c8320f1e75",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# filter_relabel_fasta.py\n",
    "# Usage: python filter_relabel_fasta.py refs.csv input.fasta output.fasta\n",
    "\n",
    "import sys\n",
    "import csv\n",
    "from Bio import SeqIO\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "71ee526e-b19c-4a91-b147-b5d111660a82",
   "metadata": {},
   "outputs": [],
   "source": [
    "csv_path = \"/Volumes/Samsung_1TB/Zooplankton/Metagenomics/DBs/DB_wrangling/speciesIDs_withOutgroups.csv\"\n",
    "fasta_in  = \"/Volumes/Samsung_1TB/Zooplankton/Metagenomics/DBs/BOLD_pull_direct/COIonly_BOLD_Public.03-Oct-2025.fasta\"\n",
    "fasta_out = \"/Volumes/Samsung_1TB/Zooplankton/Metagenomics/DBs/BOLD_pull_direct/dada_addSpecies_withOutgroups.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "172069ac-b276-40c1-98c2-5f6aaa772813",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_col = \"ID\"\n",
    "cols = [\"Kingdom\",\"Phylum\",\"Class\",\"Order\",\"Family\",\"Genus\", \"Species\"]  # levels 1..6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b2a2cec1-0a87-485d-b521-46e07128e727",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_header = {}\n",
    "with open(csv_path, newline='') as fh:\n",
    "    rdr = csv.DictReader(fh)\n",
    "    for row in rdr:\n",
    "        rid = row.get(id_col)\n",
    "        if not rid:\n",
    "            continue\n",
    "        # build Level1;Level2;...;Level6; string (coalesce missing to empty)\n",
    "        header = \";\".join((row.get(c,\"\").strip() for c in cols)) + \";\"\n",
    "        map_header[rid.strip()] = header\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "27120da2-7b29-4d69-9215-816b18a66093",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To build assign species file, build line differently\n",
    "\n",
    "id_col = \"ID\"\n",
    "cols = [\"Species\"]\n",
    "map_header = {}\n",
    "\n",
    "with open(csv_path, newline='') as fh:\n",
    "    rdr = csv.DictReader(fh)\n",
    "    for row in rdr:\n",
    "        rid = row.get(id_col)\n",
    "        if not rid:\n",
    "            continue\n",
    "        header = \" \".join((row.get(c,\"\").strip() for c in cols))\n",
    "        map_header[rid.strip()] = header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "73e6fb2c-d652-43c6-a203-b47a9fc1e739",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stream FASTA and write matches\n",
    "with open(fasta_out, \"w\") as outfh:\n",
    "    for rec in SeqIO.parse(fasta_in, \"fasta\"):\n",
    "        # rec.description contains the full header line (without leading '>')\n",
    "        # the first '|' field is the ID\n",
    "        fasta_id = rec.description.split(\"|\", 1)[0]\n",
    "       # print(fasta_id)\n",
    "        if fasta_id in map_header:\n",
    "            rec.id = (fasta_id + \" \" + map_header[fasta_id]) # SeqIO will write >rec.id\n",
    "            rec.description = \"\"             # avoid adding original desc\n",
    "            SeqIO.write(rec, outfh, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5b71d912-a93f-4596-8a81-0eeaf1b627f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Animalia;Arthropoda;Malacostraca;Isopoda;Chaetiliidae;Glyptonotus;Glyptonotus antarcticus;'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To build assignSpecies file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "8d36cb62-d612-45e2-b417-5b25b1ba2371",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "236741"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(map_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb4fcd88-514a-45d1-b2ef-a073f677324f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
